### PR TITLE
Couchbase FullTextIndex Capability

### PR DESCRIPTION
**Pull request by sandrogiacom**
#811  - Enhancement: Create CouchbaseContainer with FullTextIndex

### CODE REVIEW

**Comment by daschl** [comment]
Does this still apply after #830 @kaidowei ?

**Comment by kiview** [comment]
One more ping to @kaidowei and @sandrogiacom, does it still apply?

**Comment by kaidowei** [comment]
@kiview yes it does. :)

**Comment by sandrogiacom** [comment]
> One more ping to @kaidowei and @sandrogiacom, does it still apply?
Hi. I think I need to review and correct conflicts.

**Comment by sandrogiacom** [comment]
@kiview merge ok. Please review this. tks

**Comment by kiview** [comment]
Oh no, we have the flaky Couchbase test again :slightly_frowning_face:

**Comment by sandrogiacom** [comment]
> Oh no, we have the flaky Couchbase test again

Yes, works only 5.1.1 version :)

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 1]
```diff
@@ -271,14 +276,36 @@ public void createBucket(BucketSettings bucketSetting, UserSettings userSettings
...
+        try {
+            Path path = Paths.get(CouchbaseContainer.class.getClassLoader()
+                .getResource(ftsName + ".json").toURI());
+            Stream<String> lines = Files.lines(path);
+            lines.forEach(line -> fullTextIndex.append(line));
+            lines.close();
+
+            callCouchbaseRestAPI(fullTextIndexURL, fullTextIndex.toString(),
+                "PUT", "application/json");
+        } catch (Exception e) {
```
Comment (kiview) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java]: Can we catch more specific exceptions? E.g. when opening the file.

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 2]
```diff
@@ -271,14 +276,36 @@ public void createBucket(BucketSettings bucketSetting, UserSettings userSettings
         }
     }
 
+    private void createFullTextIndex() {
+        String urlBaseFts = String.format("http://%s:%s", getContainerIpAddress(), getMappedPort(FTS));
+        String fullTextIndexURL = urlBaseFts + "/api/index/" + ftsName;
+        StringBuilder fullTextIndex = new StringBuilder();
+        try {
+            Path path = Paths.get(CouchbaseContainer.class.getClassLoader()
+                .getResource(ftsName + ".json").toURI());
```
Comment (kiview) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java]: Should we really append the `.json` file extension here? Maybe more intuitive for the user to just specify the file name.

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 3]
```diff
@@ -380,6 +410,12 @@ public CouchbaseContainer withFts(boolean fts) {
         return self();
     }
 
+    public CouchbaseContainer withNewFts(String newFts) {
```
Comment (kiview) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java]: The way it is handled, it propose a method name like `withFtsFromClasspath`.
Also why `withNewFts` and not `withFts`?

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 4]
```diff
@@ -322,6 +349,9 @@ protected void containerIsStarted(InspectContainerResponse containerInfo) {
                 createBucket(bucket.getBucketSettings(), bucket.getUserSettings(), primaryIndex);
             }
         }
+        if (fts && ftsName != null && !ftsName.isEmpty()) {
```
Comment (kiview) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java]: I don't think the guard is needed like this. IMO checking `ftsName != null` is basically fine here.
If you want to validate correct use of the setter, probably checking for empty String would make sense in the setter, although I'm not sure how we handle in other containers.

**Comment by sandrogiacom** [comment]
I believe we can close this Issue. It is outdated. If necessary, I will open another PR. Thanks

**Pull request closed without merging**